mumez/microverse
# Feature/localization trial 2

This is the second trial of adding String localization mechanism. Please let me know what you think.

# Features
- Localizable strings are managed by text domain.
- Default domain is implicitly set based on the current behavior directory.
- You can explicitly set domain from a card 'textDomain' property.
- Localization target language will be detected by 'navigator.language', but you can also specify the target by the 'lang' url parameter.
- Translations are fall back to 'EN' if no corresponding translation is available.
- Localizable strings are dynamically loaded via fetch API.

# To test
Please access to 
http://localhost:9684/?world=localization-test
http://localhost:9684/?world=localization-test&lang=fr
http://localhost:9684/?world=localization-test&lang=ja
http://localhost:9684/?world=localization-test&lang=ja-JP (fallback to ja)
http://localhost:9684/?world=localization-test&lang=pt (fallback to en)

You will see two cards. The left card is translated by a dictionary of implicit text domain. The right card is translated using an explicitly set dictionary for "my.domain".

# Screenshots
![2023-02-10 21_48_01-_world=localization-test q=1967kt8umq#pw=or6IMdtgCM8D3IyP2Tj0iw](https://user-images.githubusercontent.com/705350/218107878-59840931-b0e2-4d9d-a204-a303e277fff8.png)
![2023-02-10 21_49_51-_world=localization-test lang=fr q=1967kt8umq#pw=or6IMdtgCM8D3IyP2Tj0iw](https://user-images.githubusercontent.com/705350/218107895-e3e7750a-b6e4-4f0b-965c-97a3e4c3259a.png)

With an explicit text domain:
![2023-02-12 00_01_56-_world=localization-test q=jvj05xsjpi#pw=ZsXIMDogJUtQHbnp6Fu03A](https://user-images.githubusercontent.com/705350/218266656-885f1280-8d45-45c0-bf23-885bbcda8a52.png)

# Notes
- Current default locales directory is `assets/locales`. Just `locales` would be better.
- Directory structure is `locales/${domain}/${lang}.json`. This is useful for packaging localization files by domain. Another option is `locales/${lang}/${domain}.json`.
- Localization file format is a plain json. Better to change to .po or .tsv format.
